// ionic-angular/tests/components/toast/toast.spec.js
"use strict";
var src_1 = require('../../../../src');
function run() {
    describe('Toast', function () {
        describe('create', function () {
            it('should create toast with close button', function () {
                var toast = src_1.Toast.create({
                    message: 'Please Wait...',
                    showCloseButton: true
                });
                expect(toast.data.position).toEqual('bottom');
                expect(toast.data.message).toEqual('Please Wait...');
                expect(toast.data.showCloseButton).toEqual(true);
            });
            it('should create toast with position top', function () {
                var toast = src_1.Toast.create({
                    message: 'Please Wait...',
                    position: 'top'
                });
                expect(toast.data.position).toEqual('top');
            });
            it('should create toast with position middle', function () {
                var toast = src_1.Toast.create({
                    message: 'Please Wait...',
                    position: 'middle'
                });
                expect(toast.data.position).toEqual('middle');
            });
            it('should create toast with position bottom', function () {
                var toast = src_1.Toast.create({
                    message: 'Please Wait...',
                    position: 'bottom'
                });
                expect(toast.data.position).toEqual('bottom');
            });
            it('should set a duration', function () {
                var toast = src_1.Toast.create({
                    message: 'Please Wait...',
                    duration: 3000
                });
                expect(toast.data.duration).toEqual(3000);
            });
        });
    });
}
exports.run = run;
