// ionic-angular/esm/components/item/item-reorder.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var __decorate = this && this.__decorate || function (decorators, target, key, desc) {
    var c = arguments.length,
        r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
        d;
    if ((typeof Reflect === "undefined" ? "undefined" : _typeof(Reflect)) === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) {
        if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    }return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = this && this.__metadata || function (k, v) {
    if ((typeof Reflect === "undefined" ? "undefined" : _typeof(Reflect)) === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = this && this.__param || function (paramIndex, decorator) {
    return function (target, key) {
        decorator(target, key, paramIndex);
    };
};
import { Component, Directive, ElementRef, EventEmitter, forwardRef, Input, NgZone, Renderer, Inject, Optional, Output } from '@angular/core';
import { Content } from '../content/content';
import { CSS } from '../../util/dom';
import { Item } from './item';
import { ItemReorderGesture } from '../item/item-reorder-gesture';
import { isTrueProperty } from '../../util/util';
/**
 * @private
 */
export var Reorder = function () {
    function Reorder(elementRef, _rendered, _zone, _content) {
        _classCallCheck(this, Reorder);

        this._rendered = _rendered;
        this._zone = _zone;
        this._content = _content;
        this._enableReorder = false;
        this._lastToIndex = -1;
        this.ionItemReorder = new EventEmitter();
        this._element = elementRef.nativeElement;
    }
    /**
     * @private
     */


    _createClass(Reorder, [{
        key: "ngOnDestroy",
        value: function ngOnDestroy() {
            this._element = null;
            this._reorderGesture && this._reorderGesture.destroy();
        }
    }, {
        key: "reorderStart",

        /**
         * @private
         */
        value: function reorderStart() {
            var children = this._element.children;
            var len = children.length;
            this.setCssClass('reorder-active', true);
            for (var i = 0; i < len; i++) {
                children[i]['$ionIndex'] = i;
            }
        }
        /**
         * @private
         */

    }, {
        key: "reorderEmit",
        value: function reorderEmit(fromIndex, toIndex) {
            var _this = this;

            this.reorderReset();
            if (fromIndex !== toIndex) {
                this._zone.run(function () {
                    _this.ionItemReorder.emit({
                        from: fromIndex,
                        to: toIndex
                    });
                });
            }
        }
        /**
         * @private
         */

    }, {
        key: "scrollContent",
        value: function scrollContent(scroll) {
            var scrollTop = this._content.getScrollTop() + scroll;
            if (scroll !== 0) {
                this._content.scrollTo(0, scrollTop, 0);
            }
            return scrollTop;
        }
        /**
         * @private
         */

    }, {
        key: "reorderReset",
        value: function reorderReset() {
            var children = this._element.children;
            var len = children.length;
            this.setCssClass('reorder-active', false);
            var transform = CSS.transform;
            for (var i = 0; i < len; i++) {
                children[i].style[transform] = '';
            }
            this._lastToIndex = -1;
        }
        /**
         * @private
         */

    }, {
        key: "reorderMove",
        value: function reorderMove(fromIndex, toIndex, itemHeight) {
            if (this._lastToIndex === -1) {
                this._lastToIndex = fromIndex;
            }
            var lastToIndex = this._lastToIndex;
            this._lastToIndex = toIndex;
            // TODO: I think both loops can be merged into a single one
            // but I had no luck last time I tried
            /********* DOM READ ********** */
            var children = this._element.children;
            /********* DOM WRITE ********* */
            var transform = CSS.transform;
            if (toIndex >= lastToIndex) {
                for (var i = lastToIndex; i <= toIndex; i++) {
                    if (i !== fromIndex) {
                        children[i].style[transform] = i > fromIndex ? "translateY(" + -itemHeight + "px)" : '';
                    }
                }
            }
            if (toIndex <= lastToIndex) {
                for (var i = toIndex; i <= lastToIndex; i++) {
                    if (i !== fromIndex) {
                        children[i].style[transform] = i < fromIndex ? "translateY(" + itemHeight + "px)" : '';
                    }
                }
            }
        }
        /**
         * @private
         */

    }, {
        key: "setCssClass",
        value: function setCssClass(classname, add) {
            this._rendered.setElementClass(this._element, classname, add);
        }
        /**
         * @private
         */

    }, {
        key: "getNativeElement",
        value: function getNativeElement() {
            return this._element;
        }
    }, {
        key: "reorder",
        get: function get() {
            return this._enableReorder;
        },
        set: function set(val) {
            this._enableReorder = isTrueProperty(val);
            if (!this._enableReorder) {
                this._reorderGesture && this._reorderGesture.destroy();
                this._reorderGesture = null;
            } else if (!this._reorderGesture) {
                console.debug('enableReorderItems');
                this._reorderGesture = new ItemReorderGesture(this);
            }
        }
    }]);

    return Reorder;
}();
__decorate([Output(), __metadata('design:type', typeof (_a = typeof EventEmitter !== 'undefined' && EventEmitter) === 'function' && _a || Object)], Reorder.prototype, "ionItemReorder", void 0);
__decorate([Input(), __metadata('design:type', Boolean)], Reorder.prototype, "reorder", null);
Reorder = __decorate([Directive({
    selector: '[reorder]',
    host: {
        '[class.reorder-enabled]': '_enableReorder'
    }
}), __param(3, Optional()), __metadata('design:paramtypes', [typeof (_b = typeof ElementRef !== 'undefined' && ElementRef) === 'function' && _b || Object, typeof (_c = typeof Renderer !== 'undefined' && Renderer) === 'function' && _c || Object, typeof (_d = typeof NgZone !== 'undefined' && NgZone) === 'function' && _d || Object, typeof (_e = typeof Content !== 'undefined' && Content) === 'function' && _e || Object])], Reorder);
/**
 * @private
 */
export var ItemReorder = function () {
    function ItemReorder(item, elementRef) {
        _classCallCheck(this, ItemReorder);

        this.item = item;
        this.elementRef = elementRef;
    }

    _createClass(ItemReorder, [{
        key: "ngAfterContentInit",
        value: function ngAfterContentInit() {
            var item = this.item.getNativeElement();
            if (item.parentNode.nodeName === 'ION-ITEM-SLIDING') {
                this.elementRef.nativeElement['$ionReorderNode'] = item.parentNode;
            } else {
                this.elementRef.nativeElement['$ionReorderNode'] = item;
            }
        }
    }]);

    return ItemReorder;
}();
ItemReorder = __decorate([Component({
    selector: 'ion-reorder',
    template: "<ion-icon name=\"menu\"></ion-icon>"
}), __param(0, Inject(forwardRef(function () {
    return Item;
}))), __metadata('design:paramtypes', [typeof (_f = typeof Item !== 'undefined' && Item) === 'function' && _f || Object, typeof (_g = typeof ElementRef !== 'undefined' && ElementRef) === 'function' && _g || Object])], ItemReorder);
var _a, _b, _c, _d, _e, _f, _g;
