// ionic-angular/esm/config/bootstrap.js
import { bootstrap } from '@angular/platform-browser-dynamic';
import { enableProdMode, NgZone, PLATFORM_DIRECTIVES, provide } from '@angular/core';
import { HTTP_PROVIDERS } from '@angular/http';
import { App } from '../components/app/app';
import { ClickBlock } from '../util/click-block';
import { closest, nativeTimeout, nativeRaf } from '../util/dom';
import { Config } from './config';
import { Events } from '../util/events';
import { FeatureDetect } from '../util/feature-detect';
import { Form } from '../util/form';
import { IONIC_DIRECTIVES } from './directives';
import { isPresent } from '../util/util';
import { Keyboard } from '../util/keyboard';
import { MenuController } from '../components/menu/menu-controller';
import { Platform } from '../platform/platform';
import { ScrollView } from '../util/scroll-view';
import { TapClick } from '../components/tap-click/tap-click';
import { Translate } from '../translation/translate';
var _reflect = Reflect;
/**
 * @name ionicBootstrap
 * @description
 * `ionicBootstrap` allows you to bootstrap your entire application. Similar to Angular's `bootstrap`, `ionicBootstrap`
 * takes a root component in order to start the app. You can pass along any providers that you may want to inject into your
 * app as an array for the second argument. You can also pass a config object as the third argument to configure your app's settings.
 *
 * @usage
 *
 * ```ts
 * import { ionicBootstrap } from 'ionic-angular';
 * import { Component } from '@angular/core';
 *
 * @Component({
 *   templateUrl: 'build/app.html',
 * })
 * export class MyClass{}
 *
 * ionicBootstrap(MyClass, null, {tabbarPlacement: 'bottom'})
 * ```
 */
export function ionicBootstrap(appRootComponent, customProviders, config) {
    // get all Ionic Providers
    var providers = ionicProviders(customProviders, config);
    // automatically set "ion-app" selector to users root component
    addSelector(appRootComponent, 'ion-app');
    cssReady(function () {
        // call angular bootstrap
        bootstrap(appRootComponent, providers).then(function (ngComponentRef) {
            // ionic app has finished bootstrapping
            ionicPostBootstrap(ngComponentRef);
        });
    });
}
/**
 * @private
 */
export function ionicPostBootstrap(ngComponentRef) {
    var app = ngComponentRef.injector.get(App);
    app.setAppInjector(ngComponentRef.injector);
    // prepare platform ready
    var platform = ngComponentRef.injector.get(Platform);
    platform.setZone(ngComponentRef.injector.get(NgZone));
    platform.prepareReady();
    // TODO: Use PLATFORM_INITIALIZER
    ngComponentRef.injector.get(TapClick);
    return ngComponentRef;
}
var cssLoadAttempt = 0;
function cssReady(done) {
    var appEle = document.body.querySelector('ion-app');
    if (!appEle || appEle.clientHeight > 0 || cssLoadAttempt > 300) {
        done();
    } else {
        nativeRaf(function () {
            cssLoadAttempt++;
            cssReady(done);
        });
    }
}
/**
 * @private
 */
export function ionicProviders(customProviders, config) {
    // create an instance of Config
    if (!(config instanceof Config)) {
        config = new Config(config);
    }
    // enable production mode if config set to true
    if (config.getBoolean('prodMode')) {
        enableProdMode();
    }
    // create an instance of Platform
    var platform = new Platform();
    // initialize platform
    platform.setUrl(window.location.href);
    platform.setUserAgent(window.navigator.userAgent);
    platform.setNavigatorPlatform(window.navigator.platform);
    platform.load(config);
    config.setPlatform(platform);
    var clickBlock = new ClickBlock();
    var events = new Events();
    var featureDetect = new FeatureDetect();
    setupDom(window, document, config, platform, clickBlock, featureDetect);
    bindEvents(window, document, platform, events);
    var providers = [App, provide(ClickBlock, { useValue: clickBlock }), provide(Config, { useValue: config }), provide(Events, { useValue: events }), provide(FeatureDetect, { useValue: featureDetect }), Form, Keyboard, MenuController, provide(Platform, { useValue: platform }), Translate, TapClick, provide(PLATFORM_DIRECTIVES, { useValue: IONIC_DIRECTIVES, multi: true }), HTTP_PROVIDERS];
    if (isPresent(customProviders)) {
        providers.push(customProviders);
    }
    return providers;
}
function setupDom(window, document, config, platform, clickBlock, featureDetect) {
    var bodyEle = document.body;
    var mode = config.get('mode');
    // if dynamic mode links have been added the fire up the correct one
    var modeLinkAttr = mode + '-href';
    var linkEle = document.head.querySelector('link[' + modeLinkAttr + ']');
    if (linkEle) {
        var href = linkEle.getAttribute(modeLinkAttr);
        linkEle.removeAttribute(modeLinkAttr);
        linkEle.href = href;
    }
    // set the mode class name
    // ios/md/wp
    bodyEle.classList.add(mode);
    // language and direction
    platform.setDir(document.documentElement.dir, false);
    platform.setLang(document.documentElement.lang, false);
    var versions = platform.versions();
    platform.platforms().forEach(function (platformName) {
        // platform-ios
        var platformClass = 'platform-' + platformName;
        bodyEle.classList.add(platformClass);
        var platformVersion = versions[platformName];
        if (platformVersion) {
            // platform-ios9
            platformClass += platformVersion.major;
            bodyEle.classList.add(platformClass);
            // platform-ios9_3
            bodyEle.classList.add(platformClass + '_' + platformVersion.minor);
        }
    });
    // touch devices should not use :hover CSS pseudo
    // enable :hover CSS when the "hoverCSS" setting is not false
    if (config.getBoolean('hoverCSS', true) !== false) {
        bodyEle.classList.add('enable-hover');
    }
    if (config.getBoolean('clickBlock', true) !== false) {
        clickBlock.enable();
    }
    // run feature detection tests
    featureDetect.run(window, document);
}
/**
 * Bind some global events and publish on the 'app' channel
 */
function bindEvents(window, document, platform, events) {
    window.addEventListener('online', function (ev) {
        events.publish('app:online', ev);
    }, false);
    window.addEventListener('offline', function (ev) {
        events.publish('app:offline', ev);
    }, false);
    window.addEventListener('orientationchange', function (ev) {
        events.publish('app:rotated', ev);
    });
    // When that status taps, we respond
    window.addEventListener('statusTap', function (ev) {
        // TODO: Make this more better
        var el = document.elementFromPoint(platform.width() / 2, platform.height() / 2);
        if (!el) {
            return;
        }
        var content = closest(el, 'scroll-content');
        if (content) {
            var scroll = new ScrollView(content);
            scroll.scrollTo(0, 0, 300);
        }
    });
    // start listening for resizes XXms after the app starts
    nativeTimeout(function () {
        window.addEventListener('resize', function () {
            platform.windowResize();
        });
    }, 2000);
}
/**
 * @private
 */
export function addSelector(type, selector) {
    if (type) {
        var annotations = _reflect.getMetadata('annotations', type);
        if (annotations && !annotations[0].selector) {
            annotations[0].selector = selector;
            _reflect.defineMetadata('annotations', annotations, type);
        }
    }
}
