// ionic-angular/config/bootstrap.js
"use strict";
var platform_browser_dynamic_1 = require('@angular/platform-browser-dynamic');
var core_1 = require('@angular/core');
var http_1 = require('@angular/http');
var app_1 = require('../components/app/app');
var click_block_1 = require('../util/click-block');
var dom_1 = require('../util/dom');
var config_1 = require('./config');
var events_1 = require('../util/events');
var feature_detect_1 = require('../util/feature-detect');
var form_1 = require('../util/form');
var directives_1 = require('./directives');
var util_1 = require('../util/util');
var keyboard_1 = require('../util/keyboard');
var menu_controller_1 = require('../components/menu/menu-controller');
var platform_1 = require('../platform/platform');
var scroll_view_1 = require('../util/scroll-view');
var tap_click_1 = require('../components/tap-click/tap-click');
var translate_1 = require('../translation/translate');
var _reflect = Reflect;
/**
 * @name ionicBootstrap
 * @description
 * `ionicBootstrap` allows you to bootstrap your entire application. Similar to Angular's `bootstrap`, `ionicBootstrap`
 * takes a root component in order to start the app. You can pass along any providers that you may want to inject into your
 * app as an array for the second argument. You can also pass a config object as the third argument to configure your app's settings.
 *
 * @usage
 *
 * ```ts
 * import { ionicBootstrap } from 'ionic-angular';
 * import { Component } from '@angular/core';
 *
 * @Component({
 *   templateUrl: 'build/app.html',
 * })
 * export class MyClass{}
 *
 * ionicBootstrap(MyClass, null, {tabbarPlacement: 'bottom'})
 * ```
 */
function ionicBootstrap(appRootComponent, customProviders, config) {
    // get all Ionic Providers
    var providers = ionicProviders(customProviders, config);
    // automatically set "ion-app" selector to users root component
    addSelector(appRootComponent, 'ion-app');
    cssReady(function () {
        // call angular bootstrap
        platform_browser_dynamic_1.bootstrap(appRootComponent, providers).then(function (ngComponentRef) {
            // ionic app has finished bootstrapping
            ionicPostBootstrap(ngComponentRef);
        });
    });
}
exports.ionicBootstrap = ionicBootstrap;
/**
 * @private
 */
function ionicPostBootstrap(ngComponentRef) {
    var app = ngComponentRef.injector.get(app_1.App);
    app.setAppInjector(ngComponentRef.injector);
    // prepare platform ready
    var platform = ngComponentRef.injector.get(platform_1.Platform);
    platform.setZone(ngComponentRef.injector.get(core_1.NgZone));
    platform.prepareReady();
    // TODO: Use PLATFORM_INITIALIZER
    ngComponentRef.injector.get(tap_click_1.TapClick);
    return ngComponentRef;
}
exports.ionicPostBootstrap = ionicPostBootstrap;
var cssLoadAttempt = 0;
function cssReady(done) {
    var appEle = document.body.querySelector('ion-app');
    if (!appEle || appEle.clientHeight > 0 || cssLoadAttempt > 300) {
        done();
    }
    else {
        dom_1.nativeRaf(function () {
            cssLoadAttempt++;
            cssReady(done);
        });
    }
}
/**
 * @private
 */
function ionicProviders(customProviders, config) {
    // create an instance of Config
    if (!(config instanceof config_1.Config)) {
        config = new config_1.Config(config);
    }
    // enable production mode if config set to true
    if (config.getBoolean('prodMode')) {
        core_1.enableProdMode();
    }
    // create an instance of Platform
    var platform = new platform_1.Platform();
    // initialize platform
    platform.setUrl(window.location.href);
    platform.setUserAgent(window.navigator.userAgent);
    platform.setNavigatorPlatform(window.navigator.platform);
    platform.load(config);
    config.setPlatform(platform);
    var clickBlock = new click_block_1.ClickBlock();
    var events = new events_1.Events();
    var featureDetect = new feature_detect_1.FeatureDetect();
    setupDom(window, document, config, platform, clickBlock, featureDetect);
    bindEvents(window, document, platform, events);
    var providers = [
        app_1.App,
        core_1.provide(click_block_1.ClickBlock, { useValue: clickBlock }),
        core_1.provide(config_1.Config, { useValue: config }),
        core_1.provide(events_1.Events, { useValue: events }),
        core_1.provide(feature_detect_1.FeatureDetect, { useValue: featureDetect }),
        form_1.Form,
        keyboard_1.Keyboard,
        menu_controller_1.MenuController,
        core_1.provide(platform_1.Platform, { useValue: platform }),
        translate_1.Translate,
        tap_click_1.TapClick,
        core_1.provide(core_1.PLATFORM_DIRECTIVES, { useValue: directives_1.IONIC_DIRECTIVES, multi: true }),
        http_1.HTTP_PROVIDERS
    ];
    if (util_1.isPresent(customProviders)) {
        providers.push(customProviders);
    }
    return providers;
}
exports.ionicProviders = ionicProviders;
function setupDom(window, document, config, platform, clickBlock, featureDetect) {
    var bodyEle = document.body;
    var mode = config.get('mode');
    // if dynamic mode links have been added the fire up the correct one
    var modeLinkAttr = mode + '-href';
    var linkEle = document.head.querySelector('link[' + modeLinkAttr + ']');
    if (linkEle) {
        var href = linkEle.getAttribute(modeLinkAttr);
        linkEle.removeAttribute(modeLinkAttr);
        linkEle.href = href;
    }
    // set the mode class name
    // ios/md/wp
    bodyEle.classList.add(mode);
    // language and direction
    platform.setDir(document.documentElement.dir, false);
    platform.setLang(document.documentElement.lang, false);
    var versions = platform.versions();
    platform.platforms().forEach(function (platformName) {
        // platform-ios
        var platformClass = 'platform-' + platformName;
        bodyEle.classList.add(platformClass);
        var platformVersion = versions[platformName];
        if (platformVersion) {
            // platform-ios9
            platformClass += platformVersion.major;
            bodyEle.classList.add(platformClass);
            // platform-ios9_3
            bodyEle.classList.add(platformClass + '_' + platformVersion.minor);
        }
    });
    // touch devices should not use :hover CSS pseudo
    // enable :hover CSS when the "hoverCSS" setting is not false
    if (config.getBoolean('hoverCSS', true) !== false) {
        bodyEle.classList.add('enable-hover');
    }
    if (config.getBoolean('clickBlock', true) !== false) {
        clickBlock.enable();
    }
    // run feature detection tests
    featureDetect.run(window, document);
}
/**
 * Bind some global events and publish on the 'app' channel
 */
function bindEvents(window, document, platform, events) {
    window.addEventListener('online', function (ev) {
        events.publish('app:online', ev);
    }, false);
    window.addEventListener('offline', function (ev) {
        events.publish('app:offline', ev);
    }, false);
    window.addEventListener('orientationchange', function (ev) {
        events.publish('app:rotated', ev);
    });
    // When that status taps, we respond
    window.addEventListener('statusTap', function (ev) {
        // TODO: Make this more better
        var el = document.elementFromPoint(platform.width() / 2, platform.height() / 2);
        if (!el) {
            return;
        }
        var content = dom_1.closest(el, 'scroll-content');
        if (content) {
            var scroll = new scroll_view_1.ScrollView(content);
            scroll.scrollTo(0, 0, 300);
        }
    });
    // start listening for resizes XXms after the app starts
    dom_1.nativeTimeout(function () {
        window.addEventListener('resize', function () {
            platform.windowResize();
        });
    }, 2000);
}
/**
 * @private
 */
function addSelector(type, selector) {
    if (type) {
        var annotations = _reflect.getMetadata('annotations', type);
        if (annotations && !annotations[0].selector) {
            annotations[0].selector = selector;
            _reflect.defineMetadata('annotations', annotations, type);
        }
    }
}
exports.addSelector = addSelector;
