// ionic-angular/esm/components/backdrop/backdrop.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var __decorate = this && this.__decorate || function (decorators, target, key, desc) {
    var c = arguments.length,
        r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
        d;
    if ((typeof Reflect === "undefined" ? "undefined" : _typeof(Reflect)) === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) {
        if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    }return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = this && this.__metadata || function (k, v) {
    if ((typeof Reflect === "undefined" ? "undefined" : _typeof(Reflect)) === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Directive, ElementRef, Input } from '@angular/core';
import { isTrueProperty } from '../../util/util';
var DISABLE_SCROLL = 'disable-scroll';
/**
 * @private
 */
var Backdrop_1 = void 0;
export var Backdrop = Backdrop_1 = function () {
    function Backdrop(elementRef) {
        _classCallCheck(this, Backdrop);

        this.elementRef = elementRef;
        this.pushed = false;
        this.disableScroll = true;
    }

    _createClass(Backdrop, [{
        key: "ngOnInit",
        value: function ngOnInit() {
            if (isTrueProperty(this.disableScroll)) {
                Backdrop_1.push();
                this.pushed = true;
            }
        }
    }, {
        key: "ngOnDestroy",
        value: function ngOnDestroy() {
            if (this.pushed) {
                Backdrop_1.pop();
                this.pushed = false;
            }
        }
    }], [{
        key: "push",
        value: function push() {
            if (this.nuBackDrops === 0) {
                console.debug('adding .disable-scroll to body');
                document.body.classList.add(DISABLE_SCROLL);
            } else {
                console.warn('several backdrops on screen? probably a bug');
            }
            this.nuBackDrops++;
        }
    }, {
        key: "pop",
        value: function pop() {
            if (this.nuBackDrops === 0) {
                console.error('pop requires a push');
                return;
            }
            this.nuBackDrops--;
            if (this.nuBackDrops === 0) {
                console.debug('removing .disable-scroll from body');
                document.body.classList.remove(DISABLE_SCROLL);
            }
        }
    }]);

    return Backdrop;
}();
Backdrop.nuBackDrops = 0;
__decorate([Input(), __metadata('design:type', Object)], Backdrop.prototype, "disableScroll", void 0);
Backdrop = Backdrop_1 = __decorate([Directive({
    selector: 'ion-backdrop',
    host: {
        'role': 'presentation',
        'tappable': '',
        'disable-activated': ''
    }
}), __metadata('design:paramtypes', [typeof (_a = typeof ElementRef !== 'undefined' && ElementRef) === 'function' && _a || Object])], Backdrop);
var _a;
