// ionic-angular/esm/util/ui-event-manager.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * @private
 */
export var PointerEvents = function () {
    function PointerEvents(ele, pointerDown, pointerMove, pointerUp, zone, option) {
        var _this = this;

        _classCallCheck(this, PointerEvents);

        this.ele = ele;
        this.pointerDown = pointerDown;
        this.pointerMove = pointerMove;
        this.pointerUp = pointerUp;
        this.zone = zone;
        this.option = option;
        this.rmTouchStart = null;
        this.rmTouchMove = null;
        this.rmTouchEnd = null;
        this.rmMouseStart = null;
        this.rmMouseMove = null;
        this.rmMouseUp = null;
        this.lastTouchEvent = 0;
        this.mouseWait = 2 * 1000;
        this.rmTouchStart = listenEvent(ele, 'touchstart', zone, option, function (ev) {
            return _this.handleTouchStart(ev);
        });
        this.rmMouseStart = listenEvent(ele, 'mousedown', zone, option, function (ev) {
            return _this.handleMouseDown(ev);
        });
    }

    _createClass(PointerEvents, [{
        key: 'handleTouchStart',
        value: function handleTouchStart(ev) {
            var _this2 = this;

            this.lastTouchEvent = Date.now() + this.mouseWait;
            if (!this.pointerDown(ev)) {
                return;
            }
            if (!this.rmTouchMove) {
                this.rmTouchMove = listenEvent(this.ele, 'touchmove', this.zone, this.option, this.pointerMove);
            }
            if (!this.rmTouchEnd) {
                this.rmTouchEnd = listenEvent(this.ele, 'touchend', this.zone, this.option, function (ev) {
                    return _this2.handleTouchEnd(ev);
                });
            }
        }
    }, {
        key: 'handleMouseDown',
        value: function handleMouseDown(ev) {
            var _this3 = this;

            if (this.lastTouchEvent > Date.now()) {
                console.debug('mousedown event dropped because of previous touch');
                return;
            }
            if (!this.pointerDown(ev)) {
                return;
            }
            if (!this.rmMouseMove) {
                this.rmMouseMove = listenEvent(window, 'mousemove', this.zone, this.option, this.pointerMove);
            }
            if (!this.rmMouseUp) {
                this.rmMouseUp = listenEvent(window, 'mouseup', this.zone, this.option, function (ev) {
                    return _this3.handleMouseUp(ev);
                });
            }
        }
    }, {
        key: 'handleTouchEnd',
        value: function handleTouchEnd(ev) {
            this.rmTouchMove && this.rmTouchMove();
            this.rmTouchMove = null;
            this.rmTouchEnd && this.rmTouchEnd();
            this.rmTouchEnd = null;
            this.pointerUp(ev);
        }
    }, {
        key: 'handleMouseUp',
        value: function handleMouseUp(ev) {
            this.rmMouseMove && this.rmMouseMove();
            this.rmMouseMove = null;
            this.rmMouseUp && this.rmMouseUp();
            this.rmMouseUp = null;
            this.pointerUp(ev);
        }
    }, {
        key: 'stop',
        value: function stop() {
            this.rmTouchMove && this.rmTouchMove();
            this.rmTouchEnd && this.rmTouchEnd();
            this.rmTouchMove = null;
            this.rmTouchEnd = null;
            this.rmMouseMove && this.rmMouseMove();
            this.rmMouseUp && this.rmMouseUp();
            this.rmMouseMove = null;
            this.rmMouseUp = null;
        }
    }, {
        key: 'destroy',
        value: function destroy() {
            this.rmTouchStart && this.rmTouchStart();
            this.rmTouchStart = null;
            this.rmMouseStart && this.rmMouseStart();
            this.rmMouseStart = null;
            this.stop();
            this.pointerDown = null;
            this.pointerMove = null;
            this.pointerUp = null;
            this.ele = null;
        }
    }]);

    return PointerEvents;
}();
/**
 * @private
 */
export var UIEventManager = function () {
    function UIEventManager() {
        var zoneWrapped = arguments.length <= 0 || arguments[0] === undefined ? true : arguments[0];

        _classCallCheck(this, UIEventManager);

        this.zoneWrapped = zoneWrapped;
        this.events = [];
    }

    _createClass(UIEventManager, [{
        key: 'listenRef',
        value: function listenRef(ref, eventName, callback, option) {
            return this.listen(ref.nativeElement, eventName, callback, option);
        }
    }, {
        key: 'pointerEventsRef',
        value: function pointerEventsRef(ref, pointerStart, pointerMove, pointerEnd, option) {
            return this.pointerEvents(ref.nativeElement, pointerStart, pointerMove, pointerEnd, option);
        }
    }, {
        key: 'pointerEvents',
        value: function pointerEvents(element, pointerDown, pointerMove, pointerUp) {
            var option = arguments.length <= 4 || arguments[4] === undefined ? false : arguments[4];

            if (!element) {
                return;
            }
            var submanager = new PointerEvents(element, pointerDown, pointerMove, pointerUp, this.zoneWrapped, option);
            var removeFunc = function removeFunc() {
                return submanager.destroy();
            };
            this.events.push(removeFunc);
            return submanager;
        }
    }, {
        key: 'listen',
        value: function listen(element, eventName, callback) {
            var option = arguments.length <= 3 || arguments[3] === undefined ? false : arguments[3];

            if (!element) {
                return;
            }
            var removeFunc = listenEvent(element, eventName, this.zoneWrapped, option, callback);
            this.events.push(removeFunc);
            return removeFunc;
        }
    }, {
        key: 'unlistenAll',
        value: function unlistenAll() {
            var _iteratorNormalCompletion = true;
            var _didIteratorError = false;
            var _iteratorError = undefined;

            try {
                for (var _iterator = this.events[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                    var event = _step.value;

                    event();
                }
            } catch (err) {
                _didIteratorError = true;
                _iteratorError = err;
            } finally {
                try {
                    if (!_iteratorNormalCompletion && _iterator.return) {
                        _iterator.return();
                    }
                } finally {
                    if (_didIteratorError) {
                        throw _iteratorError;
                    }
                }
            }

            this.events.length = 0;
        }
    }]);

    return UIEventManager;
}();
function listenEvent(ele, eventName, zoneWrapped, option, callback) {
    var rawEvent = '__zone_symbol__addEventListener' in ele && !zoneWrapped;
    if (rawEvent) {
        ele.__zone_symbol__addEventListener(eventName, callback, option);
        return function () {
            return ele.__zone_symbol__removeEventListener(eventName, callback);
        };
    } else {
        ele.addEventListener(eventName, callback, option);
        return function () {
            return ele.removeEventListener(eventName, callback);
        };
    }
}
